xchange-stream-gemini-v2/src/main/java/info/bitrich/xchangestream/gemini/GeminiStreamingMarketDataService.java

```diff
@@ -0,0 +1,72 @@
+package info.bitrich.xchangestream.gemini;
+
+import com.google.common.base.MoreObjects;
+import info.bitrich.xchangestream.core.StreamingMarketDataService;
+import io.reactivex.Observable;
+import org.knowm.xchange.currency.CurrencyPair;
+import org.knowm.xchange.dto.marketdata.OrderBook;
+import org.knowm.xchange.dto.marketdata.Ticker;
+import org.knowm.xchange.dto.marketdata.Trade;
+import org.knowm.xchange.exceptions.NotYetImplementedForExchangeException;
+
+import java.math.BigDecimal;
+import java.util.List;
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Adapted from V1 by Max Gao on 01-09-2021
+ */
+public class GeminiStreamingMarketDataService implements StreamingMarketDataService {
+    private static final String L2_UPDATES = "l2_updates";
+    private final GeminiStreamingService service;
+
+    private final Map<CurrencyPair, SortedMap<BigDecimal, BigDecimal>> bids =
+            new ConcurrentHashMap<>();
+    private final Map<CurrencyPair, SortedMap<BigDecimal, BigDecimal>> asks =
+            new ConcurrentHashMap<>();
+
+    public GeminiStreamingMarketDataService(GeminiStreamingService service) {
+        this.service = service;
+    }
+
+    private boolean containsPair(List<CurrencyPair> pairs, CurrencyPair pair) {
+        for (CurrencyPair item : pairs) {
```

Comment:
can we remove this method as per previous comment

Reply:
Done!